core/silkworm/trie/prefix_set.cpp

```diff
@@ -17,47 +17,73 @@
 #include "prefix_set.hpp"
 
 #include <algorithm>
-#include <cassert>
 
 #include <silkworm/common/util.hpp>
 
 namespace silkworm::trie {
 
-void PrefixSet::insert(ByteView key) {
-    keys_.emplace_back(key);
+void PrefixSet::insert(ByteView key, bool marker) { insert(Bytes(key), marker); }
+
+void PrefixSet::insert(Bytes&& key, bool marker) {
+    nibbled_keys_.emplace_back(key, marker);
     sorted_ = false;
 }
 
 bool PrefixSet::contains(ByteView prefix) {
-    if (keys_.empty()) {
+    if (nibbled_keys_.empty()) {
         return false;
     }
 
-    if (!sorted_) {
-        std::sort(keys_.begin(), keys_.end());
-        keys_.erase(std::unique(keys_.begin(), keys_.end()), keys_.end());
-        sorted_ = true;
-    }
+    // Key uniqueness and sorting
+    ensure_sorted();
 
-    // We optimize for the case when contains() inquires are made with increasing prefixes,
+    // We optimize for the case when contains() queries are issued with increasing prefixes,
     // e.g. contains("00"), contains("04"), contains("0b"), contains("0b05"), contains("0c"), contains("0f"), ...
     // instead of some random order.
-    assert(index_ < keys_.size());
-    while (index_ > 0 && keys_[index_] > prefix) {
+    while (index_ > 0 && nibbled_keys_[index_].first > prefix) {
         --index_;
     }
 
-    for (;; ++index_) {
-        if (keys_[index_].starts_with(prefix)) {
+    for (size_t max_index{nibbled_keys_.size() - 1};; ++index_) {
+        const auto& [key, _]{nibbled_keys_[index_]};
+        if (key.starts_with(prefix)) {
             return true;
         }
-        if (keys_[index_] > prefix) {
+        if (key > prefix || index_ == max_index) {
             return false;
         }
-        if (index_ == keys_.size() - 1) {
-            return false;
+    }
+}
+
+std::pair<bool, ByteView> PrefixSet::contains_and_next_marked(ByteView prefix) {
+    bool is_contained{contains(prefix)};
+    ByteView next_created{};
+
+    bool is_storage{prefix.length() >= 40};  // This is a prefix_set for storage changes
```

Comment:
`PrefixSet` is described as generic, but here the implementation is contaminated with specifics of a particular data set. If this optimization is material, let's expose 40 as a parameter of  `contains_and_next_marked` so that its user is warned about this peculiarity.

Reply:
My two cents ... PrefixSet is NOT a collection of prefixes rather a collection of nibbled hashed changed accounts or storage locations. So already the naming of the class is  imho misleading. What it provides, instead, is checking if the given nibbled prefix is present in any of the full (nibbled) keys owned.

Yes the change is material : as in PrefixSet populated with changes from HashedStorage all keys are prefixed with hashed contract plus incarnation (40 bytes) it allows the retrieval of a "created" item **only** within the boundaries of that specific contract. (to be noted that this also amends an [error present in erigon where the "nextMarkedItem" is searched regardless the prefixed trie owning it](https://github.com/ledgerwatch/erigon/blob/devel/turbo/trie/retain_list.go#L184-L191)).

IMHO I'd rather change the name of the class from `PrefixSet` to `HashedChangesSet` to be more consistent with its content and usage and, considering is used only (afaics) for IH i'd omit to add the parameter you requested.

Reply:
I respectfully disagree. `PrefixSet` does represent logically a set with the following property: If _x_ ∈ _S_ and _x_ starts with _y_, then _y_ ∈ _S_. But it doesn't have to store all the set elements to function correctly. And it doesn't matter whether the characters are nibbles (<16) or characters in the entire range [0; 256). So `PrefixSet` is a useful generic abstraction, which can be reasoned about independently from its use in Intermediate Hashes.

Reply:
I suggest to do the following: create a new class `HashedChangesSet` and move `contains_and_next_marked` there.  (`HashedChangesSet` can be a subclass of `PrefixSet` or has it as a private member.) Thus `PrefixSet` will stay generic and it'll be clearer than `contains_and_next_marked` is not generic.

Reply:
I respectfully disgree with you respectful disagreement.

> If x ∈ S and x starts with y, then y ∈ S

that is correct **BUT** is `y` being the prefix looked for.
So, imho, PrefixSet is not a "set of prefixes" rather a collection of discrete "full" items which might or might not start with given prefix. 

> But it doesn't have to store all the set elements to function correctly.

Honestly don't understand this.

Reply:
For example, let's consider the following mathematical set: {"", "a", "ab", "b", "bb", "bd"}, whose elements are all strings. It has the property that for every string in the set (e.g. "ab") all the possible prefixes (i.e. "" and "a") also belong to the set. `PrefixSet` represents such sets. With our implementation it can only store the longest possible strings (i.e. "ab", "bb", "bd"), omitting their shorter prefixes. But `contains()` still returns true for "", "a", "ab", "b", "bb", "bd" and nothing else.

As a counter-example,  set {"", "a", "ab", "c", "bb", "bd"} doesn't have the prefix property ("b" is a prefix of an element in the set, but doesn't belong to the set).

Reply:
In the set {"a", "ab", "b", "bb", "bd"} (note I've omitted empty string) `contains()` returns true for "", "a", "ab", "b", "bb", "bd" as every string `starts_with` an empty string (starts_with does a memcmp on zero bytes and is guaranteed to return 0)

In the set {"a", "ab", "bb", "bd", "c"} `contains()` still returns true for "", "a", "ab", "b", "bb", "bd" **and** also for "c"

So if we look at the assertion

> If x ∈ S and x starts with y, then y ∈ S

you mean the second set should enforce a return of `contains("b") == false` ? The `and` is not satisfied i.e. "b" !∈ S

Reply:
The second set {"", "a", "ab", "c", "bb", "bd"} was just an example of a set that doesn't have the prefix property (If _x_ ∈ _S_ and _x_ starts with _y_, then _y_ ∈ _S_) and thus not representable by `PrefixSet`.